### PR TITLE
PXC-3424: Fix error handling when donor is not able to serve SST

### PR DESCRIPTION
**Pull request by kamil-holubicki**
Enabled proper interpretation of result values returned by sst_donate_cb

### CODE REVIEW

**Comment by kamil-holubicki** [comment]
https://pxc.cd.percona.com/job/pxc-5.6-param/51/

File: galera/src/replicator_str.cpp [hunk 1]
```diff
@@ -342,6 +342,38 @@ ReplicatorSMM::donate_sst(void* const         recv_ctx,
     /* The fix to codership/galera#284 may break backward comatibility due to
      * different (now correct) interpretation of retrun value. Default to old
      * interpretation which is forward compatible with the new one. */
+    /* The fix to codership/galera#284 may break backward comatibility due to
+     * different (now correct) interpretation of retrun value. Default to old
+     * interpretation which is forward compatible with the new one. */
```
Comment (venkatesh-prasad-v) [inline galera/src/replicator_str.cpp]: 1. Duplicated comment?
2. s/comatibility/compatibility
3. s/retrun/return/

File: galera/src/replicator_str.cpp [hunk 2]
```diff
@@ -342,6 +342,38 @@ ReplicatorSMM::donate_sst(void* const         recv_ctx,
...
+
+    /* Backward compatibility means: WSREP API v25 + API client
+     * (sst_donate_cb) that returns positive values in case of success. 
+     * Before commit 3fd754250bc1ef4e0459848d4a390609f311c45a (galera),
+     * results >= 0 returned by API client were considered to be success.
+     * Before commit 6d1e77ad1399e9558462f56fa9e36a3217622bb1 (wsrep client),
```
Comment (venkatesh-prasad-v) [inline galera/src/replicator_str.cpp]: Commit `6d1e77ad13` doesn't exist in wsrep-API repo. Please replace `wsrep client` with `pxc` 
```suggestion
     * Before commit 6d1e77ad1399e9558462f56fa9e36a3217622bb1 (pxc), 
```

Also remove all the trailing white spaces in the newly added comment block.

File: galera/src/replicator_str.cpp (hunk 1 above)
Comment (venkatesh-prasad-v) [reply galera/src/replicator_str.cpp]: Now that we are changing the default behavior with this fix, the last sentence in the comment becomes invalid from here after. I think the last sentence can be removed.

File: galera/src/replicator_str.cpp [hunk 3]
```diff
@@ -342,6 +342,38 @@ ReplicatorSMM::donate_sst(void* const         recv_ctx,
...
+     * Before commit 6d1e77ad1399e9558462f56fa9e36a3217622bb1 (wsrep client), 
+     * wsrep_sst_donate_cb() returned WSREP_CB_FAILURE (1) also in case of
+     * success.
+     * So both above fit.
+     * 
+     * Commit  3fd754250bc1ef4e0459848d4a390609f311c45a (galera) fixed
+     * handling of errors returned by API client. Values > 0
+     * (WSREP_CB_FAILURE) are considered to be failure and 0 
+     * (WSREP_CB_SUCCESS) is considered to be success.
+     * Commit 6d1e77ad1399e9558462f56fa9e36a3217622bb1 (wsrep client) fixed
```
Comment (venkatesh-prasad-v) [inline galera/src/replicator_str.cpp]: Same as above.
```suggestion
     * Commit 6d1e77ad1399e9558462f56fa9e36a3217622bb1 (pxc) fixed
```

File: galera/src/replicator_str.cpp [hunk 4]
```diff
@@ -342,6 +342,38 @@ ReplicatorSMM::donate_sst(void* const         recv_ctx,
...
+     * WSREP_CB_SUCCESS(0) or WSREP_CB_FAILURE (1).
+     * So both above fit again.
+     * 
+     * However, everything happened without increasing wsrep API version.
+     * If we use client without the fix and galera with fix, both don't fit 
+     * as client returns 1 (WSREP_CB_FAILURE) in case of success, and galera
+     * considers it as failure.
+     * 
+     * Then we bumped wsrep API version 
+     * (commit 89b9ad07dc92eae89b6d3746e40f6044da024ccd (galera))
```
Comment (venkatesh-prasad-v) [inline galera/src/replicator_str.cpp]: This made me think that it was fixed in galera repo, which is wrong. Probably is better to change it to
```suggestion
     * (commit 89b9ad07dc92eae89b6d3746e40f6044da024ccd (wsrep-API))
```

File: galera/src/replicator_str.cpp [hunk 5]
```diff
@@ -342,6 +342,38 @@ ReplicatorSMM::donate_sst(void* const         recv_ctx,
...
+     * return code returned by wsrep_sst_donate_cb(). It returns 
+     * WSREP_CB_SUCCESS(0) or WSREP_CB_FAILURE (1).
+     * So both above fit again.
+     * 
+     * However, everything happened without increasing wsrep API version.
+     * If we use client without the fix and galera with fix, both don't fit 
+     * as client returns 1 (WSREP_CB_FAILURE) in case of success, and galera
+     * considers it as failure.
+     * 
+     * Then we bumped wsrep API version
```
Comment (venkatesh-prasad-v) [inline galera/src/replicator_str.cpp]: Not a strong opinion though, now that we are explaining everything, I feel it is also better to mention about galera/commit/0443a72a81ba77 which later fixed the SST return code compatibility to fill the missing gaps in the description.

**Review by venkatesh-prasad-v — COMMENTED** [review]
Hi Kamil,

Thank you for working on this and for the background.
I have a few suggestions. Please see below.

File: galera/src/replicator_str.cpp (hunk 1 above)
Comment (kamil-holubicki) [reply galera/src/replicator_str.cpp]: This comment comes from upstream. I duplicated it by omission. 
I explained what 'backward compatibility means in the new comment. I think we can leave it as it is now.

File: galera/src/replicator_str.cpp (hunk 2 above)
Comment (kamil-holubicki) [reply galera/src/replicator_str.cpp]: writing 'wsrep client' I have in mind "the one who uses Galera library. It is not necessarely PXC. From Galera point of view we know nothing about PXC. In fact the change was in Codership's mysql, not PXC. I changed it to 'api client'

File: galera/src/replicator_str.cpp (hunk 3 above)
Comment (kamil-holubicki) [reply galera/src/replicator_str.cpp]: as above :)

File: galera/src/replicator_str.cpp (hunk 4 above)
Comment (kamil-holubicki) [reply galera/src/replicator_str.cpp]: Done

File: galera/src/replicator_str.cpp (hunk 2 above)
Comment (venkatesh-prasad-v) [reply galera/src/replicator_str.cpp]: In that sense, I also agree with you.

**Review by venkatesh-prasad-v — APPROVED** [review]
Hi Kamil,

Thank you for addressing my comments and answering my questions. Approved from my side!
Please wait for approval from Zsolt or Kenn as I think they may have something to comment on.

**Pull request merged**